docs/architecture/adr-060-baseapp++.md

```diff
@@ -0,0 +1,200 @@
+In order to provide a flexible approach to meet the varying needs of application
+developers, we will define both a mempool interface and a data structure utilizing
+Golang generics, allowing developers to focus only on transaction
+ordering. Developers requiring absolute full control can implement their own
+custom mempool implementation.
+
+We define the general mempool interface as follows (subject to change):
+
+```go
+// MempoolTx we define a mempool transaction interface that is as minimal as
+// possible, only requiring applications to define the size of the transaction
+// to be used when reaping and getting the transaction itself. Interface type
+// casting can be used in the actual mempool implementation.
+type MempoolTx interface {
+	// Size returns the size of the transaction in bytes.
+	Size(codec.Codec) int
+	Tx() sdk.Tx
+}
+
+type Mempool[T MempoolTx] interface {
+	// Insert attempts to insert a MempoolTx into the mempool returning an error
+	// upon failure.
+	Insert(T) error
+	// Reap returns the next available transaction in the mempool, returning an error
+	// upon failure. The notion of 'available' or 'next' is defined by the application's
+	// mempool implementation.
+	Reap() (T, error)
+	// ReapMaxBytes performs the the same function as Reap, except it returns
+	// multiple transactions limited by their total size.
+	ReapMaxBytes(n int) ([]T, error)
+	// NumTxs returns the number of transactions currently in the mempool.
+	NumTxs() int
+}
+```
+
+We will define an implementation of `Mempool[T MempoolTx]` that will cover a
+majority of application use cases. Namely, it will prioritize transactions by
+priority and transaction sender, allowing for multiple prioritized transactions
+from the same sender. The mempool will be defined as a wrapper around a simple
+priority queue using a max binary heap, along with additional indexes/metadata
+to store senders and their nonces.
+
+Transaction reaping will essentially happen via a two-phase approach:
+
+1. Reap one or more transactions from the priority queue and collect them into 
+   one of two buffers -- _valid_ and _invalid_.
+2. For transactions that DO NOT violate the nonce validation, they are included
+   in the _valid_ buffer.
+3. For transactions that DO violate the nonce validation, they are included in
+   the _invalid_ buffer.
+4. Continue this process until the desired number of valid transactions are
+   reaped or until the mempool is empty.
+5. Provide Tendermint the list of all transactions from the _valid_ buffer.
+6. Re-insert all transactions, from both buffers, back into mempool. This is to
+   ensure we do not discard transactions from the mempool in case `ProcessProposal`
+   fails.
+
+```go
+type PriorityMempool[T MempoolTx] struct {
+	queue   *PriorityQueue[MempoolTx]
+	senders map[string][]int64
+	// ...
+}
+```
+
+Previous discussions<sup>1</sup> have come to the agreement that Tendermint will
```

Comment:
I think this is a good first step, but as an observation it seems like the responsibility of "who owns the mempool" is becoming a little fragmented between Tendermint and the Application. Maybe it makes sense in the future to give the Application even more control over what is cached, what should be gossiped, what should be rechecked, etc.

Reply:
From my perspective, the ownership of the mempool should really _not_ be Tendermint's. This data structure serves too many masters and has too many diverse requirements from users to be able to really handle many of these requirements _well_. A future iteration of the mempool, should likely leave gossiping as a responsibility of Tendermint and hand off the majority of the remainder of the functionality to the application, where more custom and specific logic can quickly be implemented and changed.

Reply:
@williambanfield That sounds pretty ideal to me. I hope to see things moving in that direction over time. Giving application developers more control over this (with sane defaults, of course) unlocks a lot of possibilities. 😄

Reply:
Working on writing up some thoughts on this at the moment, will happily notify you for feedback when that document is ready: https://github.com/tendermint/tendermint/issues/9240

Reply:
Man, I could not agree with this more! I totally think we need to hammer this point down.

💯💯💯💯